### DISCUSSION TITLE
`spawn_blocking` and `async`

### DISCUSSION THREAD

**Discussion by BratSinot** [tokio-rs/tokio#4149] [Q&A] [answered]
Greetings!

How can I `spawn_blocking` and use async code inside? I have some scheduler which looked-up in data (from start until the end of programm) and lot of tasks which modificates data too,

```rust
#[tokio::main]
async fn main() {
    let foo = Foo::spawn();

    for ecx in 0..10 {
        foo.add(ecx.to_string()).await;
        tokio::time::sleep(tokio::time::Duration::from_millis(10)).await;
    }

    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
}

struct Foo {
    data: std::sync::Arc<tokio::sync::RwLock<Vec<String>>>,
}

impl Foo {
    pub fn spawn() -> Self {
        let data: std::sync::Arc<tokio::sync::RwLock<_>> = Default::default();

        tokio::task::spawn_blocking({
            let data = data.clone();
            move || {
                tokio::runtime::Handle::current().block_on(print_loop(data));
            }
        });

        Self { data }
    }

    pub async fn add(&self, str: String) {
        self.data.write().await.push(str);
    }
}

async fn print_loop(data: std::sync::Arc<tokio::sync::RwLock<Vec<String>>>) {
    loop {
        println!("data: {:?}", data.read().await);
        tokio::time::sleep(tokio::time::Duration::from_millis(10)).await;
    }
}
```

**Comment by Noah-Kennedy** [comment]
Is there a particular reason this needs to be spawned as a blocking task? If you are spawning a task which does async stuff interleaved with some data processing, you are probably better of just spawning this as a normal task.

**Reply by BratSinot** [reply]
I want to spawn this in separate thread, because it interferes with other tasks (hyper / warp with a lot of WebSocket connections).

**Reply by Darksonn** [reply]
Do you only have a single task like this? Otherwise it would be better to create two independent runtimes and spawn the stuff on different runtimes.

**Reply by BratSinot** [reply]
> Do you only have a single task like this? Otherwise it would be better to create two independent runtimes and spawn the stuff on different runtimes.

Yeah, I have only single task like this.

**Comment by Darksonn** [comment, marked as answer]
You should simply be using `tokio::spawn` here.

**Reply by BratSinot** [reply]
Ordinary `tokio::spawn` with infinite loop is affecting on other task.

**Comment by Noah-Kennedy** [comment]
If this is causing problems when run in the same runtime as your other tasks, I would consider using multiple runtimes. That can often work better when this is the case.
